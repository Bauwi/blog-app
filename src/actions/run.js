import database from '../firebase/firebase';
import { db } from '../firebase/firebase';
import * as firebase from 'firebase';

// manage CRUD of run
export const addPostToRun = post => ({
  type: 'ADD_POST_TO_RUN',
  post
});

export const startAddPostToRun = post => (dispatch, getState) => {
  const { uid } = getState().auth;
  db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .add(post)
    .then((doc) => {
      dispatch(addPostToRun({
        DBid: doc.id,
        ...post
      }));
    });
};

export const removePostToRun = id => ({
  type: 'REMOVE_POST_TO_RUN',
  id
});

export const startRemovePostToRun = (id, DBid) => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .doc(`${DBid}`)
    .delete()
    .then(dispatch(removePostToRun(id)));
};

// set Run posts. This action is dispatched in the header component
// as it needs to be available for the run dropdown

export const runsHasErrored = bool => ({
  type: 'RUN_HAS_ERRORED',
  hasErrored: bool
});

export const runIsLoading = bool => ({
  type: 'RUN_IS_LOADING',
  isLoading: bool
});

export const setRunPosts = posts => ({
  type: 'SET_RUN_POSTS',
  posts
});

export const startSetRunPosts = resume => (dispatch, getState) => {
  dispatch(runIsLoading(true));
  const { uid } = getState().auth;
  const run = [];
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .get()
    .then((querySnapshot) => {
      querySnapshot.forEach((doc) => {
        run.push({
          DBid: doc.id,
          ...doc.data()
        });
      });
      return run;
    })
    .then(run => dispatch(setRunPosts(run)))
    .then(() => {
      const firstUnreadIndex = run.findIndex(post => post.state === 'unread');
      if (firstUnreadIndex === -1) {
        return dispatch(setCurrentPostRun(run[0].content.id));
      }
      return dispatch(setCurrentPostRun(run[firstUnreadIndex].content.id));
    })
    .then(() => dispatch(runIsLoading(false)))
    .catch(() => dispatch(runsHasErrored(true)));
};

// change the status of the past from read to unread
export const updateRunPostToAlreadyRead = id => ({
  type: 'UPDATE_RUN_POST_TO_ALREADY_READ',
  id
});

export const startUpdateRunPostToAlreadyRead = (id, DBid) => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .doc(`${DBid}`)
    .update({ state: 'read' })
    .then(() => {
      dispatch(updateRunPostToAlreadyRead(id));
    });
};

// this one is not used anymore. But it could be used to mark some popst as unread yet
export const updateRunPostToNotAlreadyRead = id => ({
  type: 'UPDATE_RUN_POST_TO_NOT_ALREADY_READ',
  id
});

// Set which post should actually be current. Run navigation depends on this.
// instead of routing, component is updated with currentPost data.
// easier way to deal with "Resume" to provide the user the first unread post in its run.
export const setCurrentPostRun = id => ({
  type: 'SET_CURRENT_POST_RUN',
  id
});

export const startSetCurrentPostRun = id => (dispatch) => {
  dispatch(runIsLoading(true));
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve();
    }, 10);
  })
    .then(() => dispatch(setCurrentPostRun(id)))
    .then(() => dispatch(runIsLoading(false)))
    .catch(() => dispatch(runsHasErrored(true)));
};

export const resetRun = () => ({
  type: 'RESET_RUN'
});

// remove all posts in tun
export const startResetRun = () => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .get()
    .then((querySnapshot) => {
      querySnapshot.forEach((doc) => {
        db
          .collection('runs')
          .doc(uid)
          .collection('UserRuns')
          .doc(doc.ref.id)
          .delete();
      });
    })
    .then(() => dispatch(resetRun()));
};

// Remove already read posts in run
export const cleanRun = () => ({
  type: 'CLEAN_RUN'
});

export const startCleanRun = alreadyRead => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .get()
    .then((querySnapShot) => {
      querySnapShot.forEach((doc) => {
        if (alreadyRead.includes(doc.ref.id)) {
          db
            .collection('runs')
            .doc(uid)
            .collection('UserRuns')
            .doc(doc.ref.id)
            .delete();
        }
      });
    })
    .then(dispatch(cleanRun()));
};
