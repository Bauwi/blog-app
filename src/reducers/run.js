const runReducerDefaultState = {
  isLoading: true, posts: [], hasErrored: false, current: {}
};

const runReducer = (state = runReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_POST_TO_RUN':
      return {
        ...state,
        posts: [...state.posts, action.post]
      };
    case 'REMOVE_POST_TO_RUN':
      return {
        ...state,
        posts: state.posts.filter(post => action.id !== post.content.id)
      };
    case 'SET_RUN_POSTS':
      return {
        ...state,
        posts: action.posts
      };
    case 'RUN_IS_LOADING':
      return {
        ...state,
        isLoading: action.isLoading
      };
    case 'RUN_HAS_ERRORED':
      return {
        ...state,
        hasErrored: action.hasErrored
      };
    case 'SET_CURRENT_POST_RUN':
      return {
        ...state,
        current: state.posts.find(post => post.content.id === action.id)
      };
    case 'UPDATE_RUN_POST_TO_ALREADY_READ':
      return {
        ...state,
        current: { ...state.current, state: 'read' },
        posts: state.posts.map((post) => {
          if (post.content.id === action.id) {
            return {
              ...post,
              state: 'read'
            };
          }
          return post;
        })
      };
    case 'UPDATE_RUN_POST_TO_NOT_ALREADY_READ':
      return {
        ...state,
        posts: state.posts.map((post) => {
          if (post.content.id === action.id) {
            return {
              ...post,
              state: 'unread'
            };
          }
          return post;
        })
      };
    case 'RESET_RUN':
      return {
        posts: []
      };
    case 'CLEAN_RUN':
      return {
        ...state,
        posts: state.posts.filter(post => post.state === 'unread')
      };
    default:
      return state;
  }
};

export default runReducer;
