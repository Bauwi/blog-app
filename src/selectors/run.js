const getRunPosts = (posts = [], run) => run.map(read => posts.find(post => post.id === read.id));
export default getRunPosts;
