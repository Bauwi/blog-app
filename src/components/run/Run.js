import React, { Component } from 'react';
import { connect } from 'react-redux';

import InlineList from '../InlineList';
import LoadingPage from '../LoadingPage';
import { startSetRunPosts } from '../../actions/run';

export class Run extends Component {
  state = {
    loading: true
  };
  componentWillMount() {
    if (this.props.run.length === 0) {
      this.props.startSetRunPosts().then(() => {
        this.setState(() => ({ loading: false }));
      });
    } else {
      this.setState(() => ({ loading: false }));
    }
  }

  render() {
    if (this.state.loading) {
      return <LoadingPage />;
    }
    return (
      <div>
        <p>Run</p>
        <InlineList posts={this.props.run} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startSetRunPosts: () => dispatch(startSetRunPosts())
});

const mapStateToProps = state => ({
  run: state.run
});

export default connect(mapStateToProps, mapDispatchToProps)(Run);
