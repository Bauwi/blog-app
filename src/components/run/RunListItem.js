import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import moment from 'moment';

import KeywordsList from '../KeywordsList';

import { setCurrentPostRun } from '../../actions/run';

export class RunListItem extends Component {
  handleItemClick = () => {
    this.props.history.push('/run/start');
    this.props.setCurrentPostRun(this.props.post.content.id);
  };

  render() {
    const {
      id,
      title,
      createdAt,
      author,
      keywords,
      authorId,
      miniCover,
      stars
    } = this.props.post.content;
    return (
      <div>
        <div className="inline-list-item">
          <header className="inline-list-item__left">
            <img src={miniCover} className="inline-list-item__cover" alt="post cover" />
          </header>
          <div className="inline-list-item__right">
            <section className="inline-list-item__header">
              <div>
                <i className="fa fa-star" /> {stars}
              </div>
              <div className="inline-list-item__header--right">
                {this.props.state === 'read' && (
                  <p>
                    <i className="fa fa-check" />
                  </p>
                )}
                <p className="inline-list-item__date">{moment(createdAt).format('MMM Do, YYYY')}</p>
              </div>
            </section>
            <div onClick={this.handleItemClick} className="inline-list-item__infos">
              <h3 className="inline-list-item__title">{title}</h3>
            </div>
            <footer className="inline-list-item__footer">
              <p className="inline-list-item__author">{author}</p>
              <KeywordsList keywords={keywords} />
            </footer>
          </div>
        </div>
      </div>
    );
  }
}

RunListItem.propTypes = {
  post: PropTypes.object.isRequired,
  state: PropTypes.string.isRequired,
  setCurrentPostRun: PropTypes.func.isRequired
};

const mapDispatchToProps = dispatch => ({
  setCurrentPostRun: id => dispatch(setCurrentPostRun(id))
});

export default withRouter(connect(undefined, mapDispatchToProps)(RunListItem));
