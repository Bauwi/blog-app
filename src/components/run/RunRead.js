import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import { notification } from 'antd';

import RunReadNav from './RunReadNav';
import RunReadPost from './RunReadPost';
import LoadingPage from '../LoadingPage';
import { startUpdateRunPostToAlreadyRead } from '../../actions/run';

export class RunRead extends Component {
  componentDidMount() {
    window.addEventListener('scroll', this.handleScroll);
  }

  componentWillUnmount() {
    window.removeEventListener('scroll', this.handleScroll);
  }

  handleScroll = event => {
    const DBid = this.props.current.DBid;
    if (window.innerHeight + window.scrollY >= document.body.scrollHeight) {
      if (this.props.current.state !== 'read') {
        const { id, title, author } = this.props.current.content;
        const openNotificationWithIcon = type => {
          notification[type]({
            message: title,
            description: `by ${author}`,
            style: {
              position: 'relative',
              top: '4rem'
            }
          });
        };
        openNotificationWithIcon('success');
        this.props.startUpdateRunPostToAlreadyRead(id, DBid);
      }
    }
  };

  render() {
    if (this.props.isLoading) {
      return <LoadingPage />;
    }

    return (
      <div>
        <RunReadPost post={this.props.current} />
        <RunReadNav currentRead={this.props.current} />
      </div>
    );
  }
}

RunRead.propTypes = {
  current: PropTypes.object,
  isLoading: PropTypes.bool.isRequired,
  startUpdateRunPostToAlreadyRead: PropTypes.func.isRequired
};

const mapDispatchToProps = dispatch => ({
  startUpdateRunPostToAlreadyRead: (id, DBid) => dispatch(startUpdateRunPostToAlreadyRead(id, DBid))
});

const mapStateToProps = state => ({
  current: state.run.current,
  isLoading: state.run.isLoading
});

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(RunRead));
