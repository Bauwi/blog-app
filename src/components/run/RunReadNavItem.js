import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { startSetCurrentPostRun } from '../../actions/run';

export class RunReadNavItem extends Component {
  constructor(props) {
    super(props);
    this.scrollToTop = this.scrollToTop.bind(this);
  }
  scrollToTop(scrollDuration) {
    const scrollHeight = window.scrollY,
      scrollStep = Math.PI / (scrollDuration / 15),
      cosParameter = scrollHeight / 2;
    let scrollCount = 0,
      scrollMargin,
      scrollInterval = setInterval(() => {
        if (window.scrollY != 0) {
          scrollCount += 1;
          scrollMargin = cosParameter - cosParameter * Math.cos(scrollCount * scrollStep);
          window.scrollTo(0, scrollHeight - scrollMargin);
        } else clearInterval(scrollInterval);
      }, 15);
  }
  handleChangeCurrent = () => {
    this.scrollToTop(1000);
    setTimeout(() => this.props.startSetCurrentPostRun(this.props.post.content.id), 1200);
  };

  render() {
    if (!this.props.post) {
      return (
        <li>
          <p />
        </li>
      );
    }

    const {
      id,
      title,
      createdAt,
      author,
      authorId,
      keywords,
      miniCover,
      stars
    } = this.props.post.content;

    return (
      <li className="post-nav-item" onClick={this.handleChangeCurrent}>
        <div className="post-nav-item__arrow">
          <i className="fa fa-angle-left angle-left" />
        </div>
        <div className="post-nav-item__infoscover__container">
          <header className="post-nav-item__left">
            <img src={miniCover} className="post-nav-item__cover" alt="post cover" />
          </header>
          <div className="post-nav-item__right">
            <div className="post-nav-item__infos">
              <h3 className="post-nav-item__title">{title}</h3>

              <div className="post-nav-item__subtitle">
                {this.props.post.state === 'read' ? (
                  <p>
                    <i className="fa fa-check" />
                  </p>
                ) : (
                  <p />
                )}
                <p className="post-nav-item__author">{author}</p>
              </div>
            </div>
          </div>
        </div>

        <div className="post-nav-item__arrow">
          <i className="fa fa-angle-right angle-right " />
        </div>
      </li>
    );
  }
}

RunReadNavItem.propTypes = {
  id: PropTypes.string,
  title: PropTypes.string,
  createdAt: PropTypes.number,
  author: PropTypes.object,
  keywords: PropTypes.string,
  miniCover: PropTypes.string,
  stars: PropTypes.number,
  startSetCurrentPostRun: PropTypes.func.isRequired
};

const mapDispatchToProps = dispatch => ({
  startSetCurrentPostRun: id => dispatch(startSetCurrentPostRun(id))
});

export default connect(undefined, mapDispatchToProps)(RunReadNavItem);
