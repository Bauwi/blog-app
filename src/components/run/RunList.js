import React, { Component } from 'react';
import PropTypes from 'prop-types';

import RunListItem from './RunListItem';

export default class RunList extends Component {
  renderList() {
    return this.props.posts.map((post) => {
      const key = post.content ? post.content.id : post.id;

      return <RunListItem key={key} post={post} state={post.state} />;
    });
  }

  render() {
    return (
      <div className="content-container run-list">
        {this.props.posts && this.props.posts.length !== 0 ? (
          this.renderList()
        ) : (
          <div className="run-list__empty">
            <p>Add a post to your run to start a session</p>
          </div>
        )}
      </div>
    );
  }
}

RunList.propTypes = {
  posts: PropTypes.arrayOf(PropTypes.object)
};

RunList.defaultProps = {
  posts: []
};
