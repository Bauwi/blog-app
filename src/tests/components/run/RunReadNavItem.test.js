import React from 'react';
import { shallow } from 'enzyme';
import { RunReadNavItem } from '../../../components/run/RunReadNavItem';
import run from '../../fixtures/run';

test('should render RunReadNavItem component properly', () => {
  const wrapper = shallow(<RunReadNavItem post={run.posts[2]} startSetCurrentPostRun={() => {}} />);
  expect(wrapper).toMatchSnapshot();
});
test('should render RunReadNavItem properly if empty', () => {
  const wrapper = shallow(<RunReadNavItem post={false} startSetCurrentPostRun={() => {}} />);
  expect(wrapper).toMatchSnapshot();
});

test('should call startSetCurrentPostRun after 100ms', () => {
  const startSetCurrentPostRun = jest.fn();
  const wrapper = shallow(<RunReadNavItem post={run.posts[2]} startSetCurrentPostRun={startSetCurrentPostRun} />);
  wrapper.simulate('click');
  setTimeout(() => {
    expect(startSetCurrentPostRun).toHaveBeenCalled();
  }, 100);
});

test('should not call startSetCurrentPostRun after less than 100ms', () => {
  const startSetCurrentPostRun = jest.fn();
  const wrapper = shallow(<RunReadNavItem post={run.posts[2]} startSetCurrentPostRun={startSetCurrentPostRun} />);
  wrapper.simulate('click');
  setTimeout(() => {
    expect(startSetCurrentPostRun).not.toHaveBeenCalled();
  }, 50);
});

test('should call scrollToTop on item click', () => {
  const wrapper = shallow(<RunReadNavItem post={run.posts[2]} startSetCurrentPostRun={() => {}} />);
  const scrollToTop = jest.spyOn(wrapper.instance(), 'scrollToTop');
  wrapper.update();
  wrapper.simulate('click');
  expect(scrollToTop).toHaveBeenCalled();
});
