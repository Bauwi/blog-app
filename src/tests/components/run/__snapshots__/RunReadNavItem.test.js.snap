// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render RunReadNavItem component properly 1`] = `
<li
  className="post-nav-item"
  onClick={[Function]}
>
  <div
    className="post-nav-item__arrow"
  >
    <i
      className="fa fa-angle-left angle-left"
    />
  </div>
  <div
    className="post-nav-item__infoscover__container"
  >
    <header
      className="post-nav-item__left"
    >
      <img
        alt="post cover"
        className="post-nav-item__cover"
        src="https://www.w3schools.com/w3css/img_fjords.jpg"
      />
    </header>
    <div
      className="post-nav-item__right"
    >
      <div
        className="post-nav-item__infos"
      >
        <h3
          className="post-nav-item__title"
        >
          Post 3 in Run
        </h3>
        <div
          className="post-nav-item__subtitle"
        >
          <p>
            <i
              className="fa fa-check"
            />
          </p>
          <p
            className="post-nav-item__author"
          >
            author3
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    className="post-nav-item__arrow"
  >
    <i
      className="fa fa-angle-right angle-right "
    />
  </div>
</li>
`;

exports[`should render RunReadNavItem properly if empty 1`] = `
<li>
  <p />
</li>
`;
